### DISCUSSION TITLE
Tasks arbitrarily stop scheduling when a dag is in catchup

### DISCUSSION THREAD

**Discussion by argibbs** [apache/airflow#17976] [Q&A] [answered]
I'd like to make this a bug, but I fear I don't (quite) have a solidly reproducible enough example, so discussion it is.

# Apache airflow version: 2.1.3 (was fine in 2.1.2)

# Deployment details:
- Running in docker
- Our install is not huge, we have ~150 dags of varying complexity, but most have <10 tasks per dag.
- Because Of Reasons, many of our dags run with `max_active_runs=1` and `depends_on_past=True`.
- Many of our dags have schedules that fire every 30 minutes during the working day.
- We use the celery executor with redis.
- We run multiple HA schedulers (in the affected instance, just 2)

# What happened
## Problem 1 - Dags stalling in catchup
After upgrading from 2.1.2 to 2.1.3, we observed problems with dags seemingly stalling / failing to make progress. This happened only to dags that had fallen behind for some reason (e.g. a sensor looking for files that failed to turn up) and have `catchup=True`.

When these dags fell behind, because of the changes in #16401 and #16854, the scheduler continued to create and queue the overdue dag runs. Once the issue causing the delay was resolved, we then subsequently observed that after making some progress the dags would stop scheduling. Occasionally they would make more progress but it was increasingly haphazard and very random.

Here is a (slightly redacted) screenshot of a typically "stuck" dag:
![image](https://user-images.githubusercontent.com/4406514/131752279-e165f3a8-c742-40d3-8157-bd4a5e01e33c.png)

## Problem 2 - Dags deadlocking due to out-of-order dag runs + `depends_on_past=True`
I am not sure if it's the same problem or not, but we would also occasionally observe that an arbitrary dagrun from the backlog would get picked up, rather than the "next" one. This was rare, but it definitely happened a couple of times over several days. This would leave us with an isolated dag run. However, since we have `depends_on_past=True` this dag would not make any progress. But since we have `max_active_runs=1`, the blocked dagrun was the only one that could run, and nothing else would happen.

Here is another (also slightly redacted) screenshot of an isolated dag run from a large backlog:
![image](https://user-images.githubusercontent.com/4406514/131751521-82bee170-0f9b-4566-8ca0-e746a267a5cb.png)

In the screenshot above, the dag had been stuck like this for >1 hour, while many other dags had successfully completed several runs over the same interval. ~~The affected dag had not appeared in any scheduler logs.~~ Edit: I forgot to check the logs specific to the dag file (I checked the scheduler stdout and the dag file processor. 🤦 I see that around the time the screen shot above was captured, the log file contained lines like:

```
[2021-09-01 11:55:00,662] {logging_mixin.py:109} INFO - [2021-09-01 11:55:00,662] {dag.py:2306} INFO - Setting next_dagrun for [name of dag] to 2021-08-31T20:40:00+00:00
```

Note that the given timestamp above corresponds to the final queued dag run, but is way way waaaay later than any of the stalled dag runs from the screenshot...

Note that if I edited the dag file and increased max_active_runs while in the state pictured above that an extra "thread" would get created, and pick up the "next" dag run and it would slowly make progress towards the isolated dag run.

#### Notes / comments:
I have spent several days poking around in logs and playing with settings, to no avail. I have tonight rolled back to 2.1.2 and it's like night and day - dags I have spent hours coaxing into life just cruised through their backlog and are up to date.

- The scheduler logs simply make no mention of the affected dags which makes it exceptionally hard to work out what the problem might be. Normally there's _something_ in the logs that gives you a hint, but it's as if the dags don't exist.
- The dag_file_processor log shows that the dags are parsed in a timely fashion; i.e. the scheduler _knows_ about them, it's simply not queuing the tasks (or even mentioning the dags at all!)
- Pausing some stalled dags would mean that other stalled dags would start making progress, although after each dag run completed, it felt reasonably sluggish before the next run would start running.
- Flower shows that the celery cluster was very very idle; we'd have 2 or 3 slots busy in a cluster capable of handling ~50 at once.
- Many of the affected dags had a `max_active_runs=1`.

# Related issues
I am hoping this issue has already been fixed (it may be related to #14205 and the description for #17945 looks promising) but I cannot find an explicit description of the problem we're seeing so I am listing it here.

# Things I tried:
- Increasing `max_active_runs` seemed to help sometimes, but not always.
- I set the `default_pool` to something ridiculous like 15,000 (we're normally on 500) and it didn't make any difference
- I set concurrency to 1000 (we're normally on 128) and it didn't make any difference
- I increased the dag file processor timeout to 120 (made no difference, the logs suggest the processor was comfortably parsing all dags within a couple of seconds)
- I tried deleting the dags; this sometimes helped - when they came back they would often make progress for a bit before stalling.

# How to reproduce
The first step is to create a dag that has lots of catchup. e.g. something like:
<details><summary>Example dag</summary> 

```
import datetime as dt
import pendulum

from airflow import DAG
from airflow.operators.bash import BashOperator


def create_graph(dag):
    prev_task = None
    for i in range(10):
        next_task = BashOperator(
            task_id=f'simple_task_{i}',
            bash_command="sleep 10 ; echo Dag Queue issue",
            depends_on_past=True,
            dag=dag)

        if prev_task:
            prev_task >> next_task

        prev_task = next_task


def create_dag(name: str) -> DAG:
    tz_to_use = pendulum.timezone('UTC')

    default_args = {
        'owner': 'your.email.here@email.co',
        'start_date': dt.datetime(2021, 7, 1, tzinfo=tz_to_use),
        'email': ['sla_fail@some.domain.or.other'],
        'email_on_failure': False,
        'email_on_retry': False,
    }

    dag = DAG(name,
              catchup=True,
              default_args=default_args,
              max_active_runs=1,
              schedule_interval="* * * * *")

    create_graph(dag)

    return dag

dag = create_dag("airflow_2.1.3_queue_issue")
```

</details>

## Reproducing problem 1
This is why I haven't raised a bug ... I can't work out how to do it reliably. I think maybe having lots of dags like the above would do it, but I have run out of time to experiment with 2.1.3 and have had to roll back to 2.1.2 😞 

## Reproducing problem 2
This is easier (ish). I don't know how to make it reliably launch a random dag from the backlog, but I can simulate the deadlocked behaviour, albeit within an airflow instance that is already suffering from problem 1. (It may be that a completely clean airflow instance would be fine, I don't know).

Use the dag above, let it complete several runs from the backlog, then when it's partway through a run, in the gui do "Browse" -> "Dag Runs" -> and delete the last few completed runs. You should see the deleted dag runs get recreated in the queued state, and the current dag run then deadlock because of the interaction between max_active_runs and depends_on_past.

**Comment by kaxil** [comment]
That is not a BUG, the DagRun is just in queued state, not running state. `max_active_runs` is used for controlling running state of DagRuns

**Reply by kaxil** [reply]
cc @ephraimbuddy

**Reply by kaxil** [reply]
What DB are you using for Metadata DB? Postgres or MySQL?

**Reply by ephraimbuddy** [reply]
Sounds similar to this https://github.com/apache/airflow/pull/17945#issuecomment-911357411

**Reply by ashb** [reply]
17945 would fix the issue where DagRuns stop getting put in to Running state, but this issue sounds like the problem is that tasks from running dag runs don't get scheduled, which I think iwill be a separate issue :(

**Reply by argibbs** [reply]
We use postgres 13.1 - which I believe is preferred to MySql? 

There's lots of log messages in the postgres stdout along the lines of 
```

2021-09-02 15:10:22.849 UTC [15676] ERROR:  could not obtain lock on row in relation "slot_pool"

2021-09-02 15:10:22.849 UTC [15676] STATEMENT:  SELECT slot_pool.pool AS slot_pool_pool, slot_pool.slots AS slot_pool_slots 

	FROM slot_pool FOR UPDATE NOWAIT
```

But I assumed this is harmless and just a side effect of the HA exclusive locking mechanism.

**Reply by argibbs** [reply]
I don't really mind whether you classify this as a bug or not - I partly posted this so that if anyone else hits the same problem there'll be a hit in their google search (hello [xkcd 979](https://xkcd.com/979/)) - but the fact is that I took a working 2.1.2 install that's a working clone of our prod instance, did nothing more than bump the version and apply the alembic updates ... and the whole thing ground to a halt, and there wasn't really anything I could see in the logs that would explain what was wrong.

@ashb I think (hope!) my problem here is complicated by the presence of `depends_on_past=True`, and that that's why the tasks in running dags didn't get scheduled. If so, it may all be fixed by #17945.

At this point my plan is to try the next release (assuming #17945 gets approved and merged in) and see if that's cured my problem. I'm optimistic!

If not, I'll have to schedule some time to add debug log messages to my install and try to see if I can work anything else out.

**Comment by ashb** [comment, marked as answer]
Oh this is _totally_ a bug! Yes, 19745 will be in 2.2 (and 2.1.4 if we decide there is going to be another 2.1.x release)

**Reply by potiuk** [reply]
I think we already decided to have 2.1.4 (see Slack)

**Reply by argibbs** [reply]
Hi guys,

Resurrecting this briefly to say that I've tried 2.1.4, and successfully deployed it to our dev and UAT envs, with no issues. #17945 would appear to have done the trick. I've yet to do prod, but I'm a cautious fellow so it'll be a little bit longer before it sees daylight there, but I've no reason to suspect it'll have any problems.

Thanks!

**Reply by potiuk** [reply]
+1. Airflow 2.2.0 is behind the corner, (days rather than weeks) so yeah wait a bit (but no longer than that)

**Comment by potiuk** [comment]
BTW. I actually even tried to convert it to an "issue" right after you opened it @argibbs - cause the quality of the issue report was fantastic  even without ful reproduction steps!

The problem was that it was ...... too long. GitHub issue make the conversion by converting the issue content into URL to submit a new issue ! and when I clicked "Create issue from Discussion" I got the pink unicorn and "URL too long" error.

But this is definitely an ISSUE  and big thanks for being so detailed about it. Next time just open an issue if you provide that many details without bothering about complete reproduction steps! 

And it's really min-boggling BTW, how people who are so dilligent and provide so many details are also so dilligent in reading and following our comments when you create the issue. I've learned something today that might lead to rewording our issue templates a bit.

Thanks AGAIN! @argibbs !
